torch/csrc/lazy/core/dynamic_ir.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <ATen/core/symbol.h>
+
+#include <functional>
+#include <memory>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <c10/core/ScalarType.h>
+#include <torch/csrc/lazy/core/hash.h>
+#include <torch/csrc/lazy/core/ir.h>
+#include <torch/csrc/lazy/core/ir_metadata.h>
+#include <torch/csrc/lazy/ts_backend/ts_node.h>
+#include <c10/util/Flags.h>
+
+C10_DECLARE_bool(ltc_enable_dynamic_shapes);
+
+namespace torch {
+namespace lazy {
+
+/**
+ * The goal of "dynamic" Nodes is to patch a hole in our tracing.
+ * Previously, if a user called `sizes` on a Tensor, it would leak out
+ * of our tracing system, as `sizes` returns a torch.Size or an int. To
+ * prevent this from happening, we introduce DimensionNode, a new type
+ * of Node that abstracts the operation of getting the dimensions of a
+ * Tensor.
+ * 
+ * Consider the following example:
+ * ```
+ * numel = x.shape()[0] * x.shape()[1]
+ * ```
+ * 
+ * Here, `x.shape()[i]` will be a SizeNode (subclass of DimensionNode),
+ * and the multiplication of the two SizeNodes will be represented by
+ * a SizeMul (also a subclass of DimensionNode). Through this, we can
+ * prevent `numel` from being represented as a Python int and thus
+ * burned into the Graph.
+ */
+
+class TORCH_API DimensionNode : public lazy::TsNode {
```

Comment:
PT/XLA LTC relies on [`lazy::Node`](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/ir.h#L96). Do you plan to subclass `DimensionNode` from `lazy::Node` later?

Reply:
Design section [C++ Level API (PyTorch/LTC)](https://docs.google.com/document/d/1qUBwlLbDPsWKP3kaSCEgIF4hf-C4vvpA10upEvZep_s/edit#heading=h.h59w757f5joi)

Reply:
@Krovatkin what are the blockers to this exactly?
1) Node was missing shapes_ and operands_ fields, etc
2) we have this general assumption that TsNode is the base of all our nodes

(1) was fixed yesterday by @antoniojkim with #74710
(2) is still an open question for 2 reasons,
  2a - we planned to rely on TsNode to implement shape cache and keep Node simpler/unaware of 'hash_with/without_sizes
  2b - Xla would rely on having XlaNode as a base class at least for a long time, if just for migration purposes

Reply:
in other words, is it going to be OK if DiminsionNode extends from Node?  Or does that violate our design that backends can rely on all nodes deriving from {Backend}Node?